cms/lib/xblock/upstream_sync.py

```diff
@@ -0,0 +1,228 @@
+        help=(
+            "Names of the fields which have values set on the upstream block yet have been explicitly overridden "
+            "on this downstream block. Unless explicitly cleared by ther user, these overrides will persist even "
+            "when updates are synced from the upstream."
+        ),
+        hidden=True,
+        default=[],
+        enforce_type=True,
+    )
+    upstream_settings = Dict(
+        scope=Scope.settings,
+        help=(
+            "@@TODO helptext"
+        ),
+        hidden=True,
+        default={}, enforce_type=True,
+    )
+
+    def save(self, *args, **kwargs):
+        """
+        Upon save, ensure that upstream_overriden tracks all upstream-provided fields which downstream has overridden.
+
+        @@TODO use is_dirty instead of getattr for efficiency?
+        """
+        for field_name, value in self.upstream_settings.items():
+            if field_name not in self.upstream_overridden:
+                if value != getattr(self, field_name):
+                    self.upstream_overridden.append(field_name)
+        super().save()
+
+    @XBlock.handler
+    def sync_updates(self, request: Request, suffix=None) -> Response:
+        """
+        XBlock handler
+        """
+        if request.method != "POST":
+            return Response(status_code=405)
+        if not self.upstream:
+            return Response("no linked upstream", response=400)
+        upstream_info = self.get_upstream_info()
+        if upstream_info["error"]:
+            return Response(upstream_info["error"], status_code=400)
+        self.sync_from_upstream(user=request.user, apply_updates=True)
+        self.save()
+        self.runtime.modulestore.update_item(self, request.user.id)
+        return Response(json.dumps(self.get_upstream_info()), indent=4)
+
+    def sync_from_upstream(self, *, user: User, apply_updates: bool) -> None:
+        """
+        @@TODO docstring
+
+        Does NOT save the block; that is left to the caller.
+
+        Raises: InvalidKeyError, UnsupportedUpstreamKeyType, XBlockNotFoundError
+        """
+        if not self.upstream:
+            self.upstream_settings = {}
+            self.upstream_overridden = []
+            self.upstream_version = None
+            return
+        upstream_key = validate_upstream_key(self.upstream)
+        self.upstream_settings = {}
+        try:
+            upstream_block = xblock_api.load_block(upstream_key, user)
+        except NotFound as exc:
+            raise XBlockNotFoundError(
+                f"failed to load upstream ({self.upstream}) for block ({self.usage_key}) for user id={user.id}"
+            ) from exc
+        for field_name, field in upstream_block.fields.items():
+            if field.scope not in [Scope.settings, Scope.content]:
+                continue
+            value = getattr(upstream_block, field_name)
+            if field.scope == Scope.settings:
+                self.upstream_settings[field_name] = value
```

Comment:
I think we should do that thing we talked about where we have a fixed set of known overridable fields, and not generally pull from the settings scope. Otherwise we're going to make pretty much the entire VideoBlock overridable:

https://github.com/openedx/edx-platform/blob/ca46c20abb646b8b887e3c4d999f4e069f7eb871/xmodule/video_block/video_xfields.py#L33-L122

And we're going to run into that weird issue where the ProblemBlock's `markdown` field is settings scope, but changes to it try to write to the content scoped `data` field and just mess everything up:

https://github.com/openedx/edx-platform/blob/ca46c20abb646b8b887e3c4d999f4e069f7eb871/xmodule/capa_block.py#L323

Reply:
Done `get_customizable_fields`